### ISSUE TITLE
high frequency fitting problem

### ISSUE THREAD

**Issue by z1q0831** [ECSHackWeek/impedance.py#283]
hallo, i have a question about high freq fitting. I have used RL to fit but seems does not work.
the fitting of positive Zimg is not possible
can anyone maybe know why?plz
![Snipaste_2023-10-25_21-55-16](https://github.com/ECSHackWeek/impedance.py/assets/149013153/7dc2e0d3-6cc9-4a78-a924-ed1b0ef31014)
![Snipaste_2023-10-25_21-55-32](https://github.com/ECSHackWeek/impedance.py/assets/149013153/cc23de70-e8b1-4265-ac5a-65bdef6f333a)

**Comment by z1q0831** [comment]
@Yugal0

**Comment by Yugal0** [comment]
Although I will give it a try myself tomorrow, but for now I think you can try increasing the value of inductance in the initial guess.

**Comment by z1q0831** [comment]
> Although I will give it a try myself tomorrow, but for now I think you can try increasing the value of inductance in the initial guess.

thank you for the suggestion. I have tried to increase the value of inductor, but still can not get the the fitting of positive Zimag. the csv i use is from the [exampleData.csv](https://impedancepy.readthedocs.io/en/latest/_downloads/320671c0bb666e4d6ac487c9d7ff1679/exampleData.csv).

**Comment by Yugal0** [comment]
> > Although I will give it a try myself tomorrow, but for now I think you can try increasing the value of inductance in the initial guess.
> 
> thank you for the suggestion. I have tried to increase the value of inductor, but still can not get the the fitting of positive Zimag. the csv i use is from the [exampleData.csv](https://impedancepy.readthedocs.io/en/latest/_downloads/320671c0bb666e4d6ac487c9d7ff1679/exampleData.csv).

Ok I will try fitting myself.(tommorow)

**Comment by Yugal0** [comment]
What I got in the first try:

`from impedance.models.circuits import CustomCircuit`
`import numpy as np`
`import pandas as pd`
`import matplotlib.pyplot as plt`

`initial_guess=[0.01,0.01,100,0.01,0.05,100,1,1,0.1]`
`c=CustomCircuit("R0-p(R1,C1)-p(R2-Wo1,C2)-p(L1,R3)",initial_guess=initial_guess)`
`df=pd.read_csv("../exampleData.csv")`
`gthbdata=df.to_numpy()`
`datanew=np.transpose(gthbdata)`
`freq=datanew[0]`
`z_data=datanew[1]+1j*datanew[2]`
`plt.plot(np.real(z_data),-np.imag(z_data))`
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/f830b57c-c97c-4296-97a4-affbb4cb975c)

`c.fit(frequencies=freq,impedance=z_data)`
`print(c)`
`z_fit=c.predict(frequencies=freq)`
`plt.plot(np.real(z_fit),-np.imag(z_fit))`
`plt.plot(np.real(z_data),-np.imag(z_data))`
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/acfcf50a-925a-46c3-8365-ba85f431fce0)
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/597e746c-233c-4164-a316-332350f1c764)

**Comment by z1q0831** [comment]
> What I got in the first try:
> 
> `from impedance.models.circuits import CustomCircuit` `import numpy as np` `import pandas as pd` `import matplotlib.pyplot as plt`
> 
> `initial_guess=[0.01,0.01,100,0.01,0.05,100,1,1,0.1]` `c=CustomCircuit("R0-p(R1,C1)-p(R2-Wo1,C2)-p(L1,R3)",initial_guess=initial_guess)` `df=pd.read_csv("../exampleData.csv")` `gthbdata=df.to_numpy()` `datanew=np.transpose(gthbdata)` `freq=datanew[0]` `z_data=datanew[1]+1j*datanew[2]` `plt.plot(np.real(z_data),-np.imag(z_data))` ![image](https://user-images.githubusercontent.com/122299659/278745289-f830b57c-c97c-4296-97a4-affbb4cb975c.png)
> 
> `c.fit(frequencies=freq,impedance=z_data)` `print(c)` `z_fit=c.predict(frequencies=freq)` `plt.plot(np.real(z_fit),-np.imag(z_fit))` `plt.plot(np.real(z_data),-np.imag(z_data))` ![image](https://user-images.githubusercontent.com/122299659/278745592-acfcf50a-925a-46c3-8365-ba85f431fce0.png) ![image](https://user-images.githubusercontent.com/122299659/278745639-597e746c-233c-4164-a316-332350f1c764.png)

thank you for your try, so as I see in the last picture, the fitting line also can not reach the positive Zimag?

**Comment by Yugal0** [comment]
> > What I got in the first try:
> > `from impedance.models.circuits import CustomCircuit` `import numpy as np` `import pandas as pd` `import matplotlib.pyplot as plt`
> > `initial_guess=[0.01,0.01,100,0.01,0.05,100,1,1,0.1]` `c=CustomCircuit("R0-p(R1,C1)-p(R2-Wo1,C2)-p(L1,R3)",initial_guess=initial_guess)` `df=pd.read_csv("../exampleData.csv")` `gthbdata=df.to_numpy()` `datanew=np.transpose(gthbdata)` `freq=datanew[0]` `z_data=datanew[1]+1j*datanew[2]` `plt.plot(np.real(z_data),-np.imag(z_data))` ![image](https://user-images.githubusercontent.com/122299659/278745289-f830b57c-c97c-4296-97a4-affbb4cb975c.png)
> > `c.fit(frequencies=freq,impedance=z_data)` `print(c)` `z_fit=c.predict(frequencies=freq)` `plt.plot(np.real(z_fit),-np.imag(z_fit))` `plt.plot(np.real(z_data),-np.imag(z_data))` ![image](https://user-images.githubusercontent.com/122299659/278745592-acfcf50a-925a-46c3-8365-ba85f431fce0.png) ![image](https://user-images.githubusercontent.com/122299659/278745639-597e746c-233c-4164-a316-332350f1c764.png)
> 
> thank you for your try, so as I see in the last picture, the fitting line also can not reach the positive Zimag?

Yes, same problem as yours happening.
But I made another mistake, the plots are not 1:1 in aspect ratio.
And also, seems like the fit you are getting and what I am getting are different, why so? what method are you using for fitting?
YOUR FIT:
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/01632a35-329a-4d7a-a109-2b731892149e)
MY FIT: (now its 1:1)
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/0cbd519f-1cd5-4644-bce5-2933b0646df6)

**Comment by z1q0831** [comment]
> > > What I got in the first try:
> > > `from impedance.models.circuits import CustomCircuit` `import numpy as np` `import pandas as pd` `import matplotlib.pyplot as plt`
> > > `initial_guess=[0.01,0.01,100,0.01,0.05,100,1,1,0.1]` `c=CustomCircuit("R0-p(R1,C1)-p(R2-Wo1,C2)-p(L1,R3)",initial_guess=initial_guess)` `df=pd.read_csv("../exampleData.csv")` `gthbdata=df.to_numpy()` `datanew=np.transpose(gthbdata)` `freq=datanew[0]` `z_data=datanew[1]+1j*datanew[2]` `plt.plot(np.real(z_data),-np.imag(z_data))` ![image](https://user-images.githubusercontent.com/122299659/278745289-f830b57c-c97c-4296-97a4-affbb4cb975c.png)
> > > `c.fit(frequencies=freq,impedance=z_data)` `print(c)` `z_fit=c.predict(frequencies=freq)` `plt.plot(np.real(z_fit),-np.imag(z_fit))` `plt.plot(np.real(z_data),-np.imag(z_data))` ![image](https://user-images.githubusercontent.com/122299659/278745592-acfcf50a-925a-46c3-8365-ba85f431fce0.png) ![image](https://user-images.githubusercontent.com/122299659/278745639-597e746c-233c-4164-a316-332350f1c764.png)
> > 
> > 
> > thank you for your try, so as I see in the last picture, the fitting line also can not reach the positive Zimag?
> 
> Yes, same problem as yours happening. But I made another mistake, the plots are not 1:1 in aspect ratio. And also, seems like the fit you are getting and what I am getting are different, why so? what method are you using for fitting? YOUR FIT: ![image](https://user-images.githubusercontent.com/122299659/278750404-01632a35-329a-4d7a-a109-2b731892149e.png) MY FIT: (now its 1:1) ![image](https://user-images.githubusercontent.com/122299659/278750550-0cbd519f-1cd5-4644-bce5-2933b0646df6.png)

do you have the same initial guess of mine? Maybe i have add another (RC) component, because (RC) is responsible for semicircle part.

**Comment by Yugal0** [comment]
yes I put the exact same initial guess (and the same elements) as you showed in the code in your very first snapshot in the question. 

Actually just now, what I found is if I take L=100 as the initial guess the I am getting a pretty good fit 
MY FIT NOW:
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/60a86ebe-2899-41fc-a1f1-937198de2c64)
In your snapshot it was L=1, with that I was getting not a good fit (as shown in my previous comments) 
MY FIT BEFORE:
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/55da7443-0429-4a78-b329-9a3ec44fea6e)

But still how you getting good fit with L=1 as initial guess?
YOUR FIT as shown in your first comment:
![image](https://github.com/ECSHackWeek/impedance.py/assets/122299659/987b9726-b3c1-4a6d-bffa-baee7775bf50)

**Comment by z1q0831** [comment]
> yes I put the exact same initial guess (and the same elements) as you showed in the code in your very first snapshot in the question. YOUR SNAPSHOT of CODE: ![image](https://user-images.githubusercontent.com/122299659/278754568-ef82668a-f7be-4929-a549-f4b2b3f161b2.png)
> 
> Actually just now, what I found is if I take L=100 as the initial guess the I am getting a pretty good fit MY FIT NOW: ![image](https://user-images.githubusercontent.com/122299659/278754098-60a86ebe-2899-41fc-a1f1-937198de2c64.png) In your snapshot it was L=1, with that I was getting not a good fit (as shown in my previous comments) MY FIT BEFORE: ![image](https://user-images.githubusercontent.com/122299659/278754376-55da7443-0429-4a78-b329-9a3ec44fea6e.png)
> 
> But still how you getting good fit with L=1 as initial guess? YOUR FIT as shown in your first comment: ![image](https://user-images.githubusercontent.com/122299659/278754437-987b9726-b3c1-4a6d-bffa-baee7775bf50.png)

Maybe I have changed the value of L , but the problem still remains. The high frequency fitting can not be reached.

**Comment by Yugal0** [comment]
Yeah, that is the main problem. I will let you know here, If I get any solutions for this.

**Comment by z1q0831** [comment]
> Yeah, that is the main problem. I will let you know here, If I get any solutions for this.

But totally thank you for replying me! If I find any solutions, I will post here

**Comment by BGerwe** [comment]
Closing because this thread is stale. Hopefully you were able to fully resolve your problem
